### PR TITLE
Iox #521 best practice for testing

### LINKED ISSUE #521: Improve Pull-Request Process
## Brief feature description
In iceoryx are Pull-Requests often a long time open because of several inconsistencies:
- New contributors often don't know who is the committer with the best knowledge about the codebase where he is working on
- There are formal aspects expected from committer side which are not visible to the contributors (e.g. AAA Pattern in tests, which cases are needed to test? How to name tests properly?)
- The expectations regarding the scope of a PR are sometimes unclear (e.g. When is a PR too huge? What is expected on tests?)

Of course a Pull-Request is something iterative and the last thing we want is to have fast merges without reviewing it accordingly.
For more efficient Pull-Request reviews we need to improve the Contribution Process and give more information about it.

In the Contributing.md are already helpful hints listed: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
When i take a look at https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#testing i see not much info here, this can be a starting point to extend it.

Possible solutions (to be extended):
- [ ] extend the Contributor instructions with more hints
- [ ] add codeowners to iceoryx components which are automatically added to the Pull-Request. https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners
- [x] more automatic checks at Pull-Request like if all points on the PR checklist are marked
- [ ] add a PR Merge-Bot which ensures syncing with master and do automatic checks
- [x] add labels to PR template so that Contributors can set them
- [x] enable annotations for codecov.io

Maybe we can have here a starting point with a checklist (unordered and incomplete):
- [ ] build everything and execute tests: `./tools/iceoryx_build_test.sh build-strict build-all out-of-tree test clean`
- [ ] are unittests complete? [Link](https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#testing)
- [ ] documentation in doxygen complete? [Link](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/aspice_swe3_4/swe_docu_guidelines.md)
- [ ] clang-format the code?
- [ ] naming correct? See: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#naming-conventions
- [ ] fileheader correct? (company)
- [ ] use `const &` in function arguments wherever possible
- [ ] avoid using fundamental types like `int`, use fixed width integer types like `uint32_t`
- [ ] always initialize member variables and use uppercase literals if possible `uint32_t foo{0};`
- [ ] no commented code
- [ ] Are the labels in the PR properly set? (E.g. test, refactoring etc.)
- [x] ...

checklist for tests in Contributing.md(incomplete):
- [x] use AAA pattern for testcase structure https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80
- [x] use `sut` as member name for the object to test
- [x] prefer to use stack-allocated objects for testing, if only heap is possible then wrap them in smart-pointers to ensure correct cleanup and avoid memleaks
- [x] create a meaningful name for your testcase, e.g. `TEST_F(MemoryManager_test, PerformSomethingAndExpectThis)`
- [x] don't test multiple aspects in one testcase, prefer to split up in different little testcases
- [x] avoid usage of `new` in tests as much as possible because you need to `delete` it afterwards
- [x] test negative cases too
- [x] possible abstract testcases: zero, one, many, corner cases, limits
- [x] ...

Feel free to make proposals and extend the checklist.
@eclipse-iceoryx/technology-iceoryx-committers

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer

This PR tries to document the our best practices to write tests for iceoryx. Some of them were unwritten, making it hard for new contributor to write test the way we expect it.
I still have to write a small example for typed and parameterized test, but this will be more or less just a template for c&p and all the other stuff is done and can be reviewed.
Oh, and I fixed a typo from my last PR.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [x] All checkboxes in the PR checklist are checked or crossed out
1. [x] Merge

## References

- Partly closes #521

### CODE REVIEW

**Comment by dkroenke** [comment]
There should be a link from https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#testing to this document.

File: doc/website/advanced/best-practice-for-testing.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+## Mocks
+
+TODO:
+- just a few links
+
+# Conclusion
+
+- apply the AAA pattern to structure the test and check only one aspect of the code at a time
+- use the ZOMBIES principle to find sensible test cases
+- meaningful names for the tests to indicate what the test is supposed to do
```
Comment (marthtz) [inline doc/website/advanced/best-practice-for-testing.md:251]: I'd be more explicit here. The test name should indicate what's done in the test (or going in) and include the outcome of the test, e.g. "...IsSuccessful" or "...ReturnsXYZ".

File: doc/website/advanced/best-practice-for-testing.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
+# Best practice for testing
+
+This is a guide on how to write tests for iceoryx. It intends to cover the most common cases which will probably apply to 99% of the tests.
```
Comment (marthtz) [inline doc/website/advanced/best-practice-for-testing.md:3]: I'd remove `the` and write `99% of tests.`

File: doc/website/advanced/best-practice-for-testing.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+This shall not encourage to be more royalist than the king and common sense shall be applied when the guidelines don't make sense.
+But if half of the tests don't follow the guidelines, it's a clear indicator that they need to be refactored.
+
+The guide also expects some knowledge on `gtest`. At least the [Googletest Primer](https://github.com/google/googletest/blob/master/docs/primer.md) document should be read before continuing with this guide.
+
+Don't write test just for the sake of having a high coverage number.
```
Comment (marthtz) [inline doc/website/advanced/best-practice-for-testing.md:9]: Don't write test**s**...

File: doc/website/advanced/best-practice-for-testing.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    - is the copy origin unchanged
+    - etc.
+- move
+    - is the move destination object cleaning up its resources
+    - is the move origin object in a defined but unspecified state
+    - etc.
+- etc.
+
+Following [Hyrum's Law](https://www.hyrumslaw.com/) loosely, given enough users, one will find ways to use the software in a way it was never imagined.
+Therefore never underestimate the creativity of brilliancy/stupidity.
```
Comment (marthtz) [inline doc/website/advanced/best-practice-for-testing.md:57]: I'd add a comma: `Therefore,`

File: doc/website/advanced/best-practice-for-testing.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+TODO:
+- simple example which can be used as template
+- https://www.sandordargo.com/blog/2019/04/24/parameterized-testing-with-gtest
+- use `std::tie`
+
+## Mocks
+
+TODO:
+- just a few links
```
Comment (dkroenke) [inline doc/website/advanced/best-practice-for-testing.md:245]: Everything in https://github.com/google/googletest/tree/release-1.10.0/googlemock/docs

File: doc/website/advanced/best-practice-for-testing.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+A good reason to use the heap are large objects which might cause a stack overflow.
+Some operating system have a rather small stack of only a few kB, so this limit might be closer one might think.
+
+Sometimes it is necessary to access a pointer returned by a method.
+Using an `EXPECT_*` to check for a `nullptr` is not sufficient if the pointer is dereferenced afterwards.
+It might crash the test application if it indeed is a `nullptr`.
+The implementation of the test object might be broken and only after the successful test run it can be assumed to be correct.
+To prevent a crash, the `nullptr` check has to be done with an `ASSERT_*`. This gracefully aborts the current test case and continues with the next one.
+The same applies to other potential dangerous operations, like accessing the value of an `cxx::optional` or `cxx::expected`.
+This check can be omitted if another test case already ensured that the test object operates as expected.
```
Comment (elBoberido) [inline doc/website/advanced/best-practice-for-testing.md:71]: I'm not quite sure about this anymore. Let's assume the implementation is faulty. The test for that specific aspect will fail gracefully. Subsequent tests accessing this data will crash the test app. Should we state that potential dangerous behavior should always be asserted to gracefully abort a test?

File: doc/website/advanced/best-practice-for-testing.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+- use `std::tie`
+
+## Mocks
+
+TODO:
+- just a few links
+
+# Conclusion
+
+- apply the AAA pattern to structure the test and check only one aspect of the code at a time
```
Comment (dkroenke) [inline doc/website/advanced/best-practice-for-testing.md:249]: "check only one aspect of the code per testcase"

File: doc/website/advanced/best-practice-for-testing.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+These are some examples showing how to apply the ZOMBIES principle to find good test cases.
+They also exert all the good practices mentioned previously, like clear and distinct names or avoiding magic numbers.
+
+# Slightly More Advanced Topics
+
+## Typed Tests
+
+TODO:
+- simple example which can be used as template
```
Comment (dkroenke) [inline doc/website/advanced/best-practice-for-testing.md:233]: Google has a good commented example: https://github.com/google/googletest/blob/release-1.8.1/googletest/samples/sample6_unittest.cc

File: doc/website/advanced/best-practice-for-testing.md (hunk 7 above)
Comment (marthtz) [reply doc/website/advanced/best-practice-for-testing.md:249]: Maybe also add something like "once it's been tested it doesn't need to be retested again in another test" or so.

File: doc/website/advanced/best-practice-for-testing.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+O = One
+M = Many (or More complex)
+B = Boundary Behaviors
+I = Interface Definition
+E = Exercise Exceptional Behavior
+S = Simple Scenarios, Simple Solutions
+```
+
+This can be separated into **ZOM** and **BIE** with **S** bridging them together.
+**ZOM** are often simple test, like _a vector with zero items is empty_ or _a vector with one item is not empty_ or _a vector with N items has a size of N_.
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:30]: simple test**s**

File: doc/website/advanced/best-practice-for-testing.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    constexpr uint64_t EXPECTED_VALUE{0U};
+
+    SingleDigitNumber sut;
+
+    EXPECT_EQ(static_cast<uint64_t>(sut), EXPECTED_VALUE);
+}
+```
+
+The test also has a meaningful name. If this fails in the CI, it is immediately clear what is broken.
+Additionally, the tested object is called `sut`. This makes it easy to identify the actual test object.
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:174]: Maybe you could add the meaning of ``sut`` (or is this clear to everyone?)

File: doc/website/advanced/best-practice-for-testing.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+- just a few links
+
+# Conclusion
+
+- apply the AAA pattern to structure the test and check only one aspect of the code at a time
+- use the ZOMBIES principle to find sensible test cases
+- meaningful names for the tests to indicate what the test is supposed to do
+- name the test object `sut` to make it clear what object is tested
+- don't use magic numbers
+- instantiate objects on the stack or use smart pointer for large objects and avoid manual memory management with new/delete
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:254]: smart pointer**s**

File: doc/website/advanced/best-practice-for-testing.md (hunk 7 above)
Comment (elBoberido) [reply doc/website/advanced/best-practice-for-testing.md:249]: are you fine with `don't test previously tested behavior`?

**Review by dkroenke — DISMISSED** [review]

File: doc/website/advanced/best-practice-for-testing.md (hunk 1 above)
Comment (elBoberido) [reply doc/website/advanced/best-practice-for-testing.md:257]: I'm a bit hesitant because I don't want to put us in a straitjacket. How about "meaningful names for the tests to indicate what the test is supposed to do and the outcome" and adding this as examples but not limiting it to this?

File: doc/website/advanced/best-practice-for-testing.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+To avoid manual memory management with new/delete, smart pointer shall be used if possible.
+As a reminder, if a method takes a pointer to an object, this object can be instantiated on the stack and the address of this object can be passed to the method.
+A good reason to use the heap are large objects which might cause a stack overflow.
+Some operating system have a rather small stack of only a few kB, so this limit might be closer one might think.
+
+Sometimes it is necessary to access a pointer returned by a method.
+Using an `EXPECT_*` to check for a `nullptr` is not sufficient if the pointer is dereferenced afterwards.
+It might crash the test application if it indeed is a `nullptr`.
+The implementation of the test object might be broken and only after the successful test run it can be assumed to be correct.
+To prevent a crash, the `nullptr` check has to be done with an `ASSERT_*`. This gracefully aborts the current test case and continues with the next one.
```
Comment (MatthiasKillat) [inline doc/website/advanced/best-practice-for-testing.md:69]: In general ASSERT is to be used if te test cannot continue if the condition is not true (e.g. nullptr, empty optional, non-existing element in an array and similar cases)

File: doc/website/advanced/best-practice-for-testing.md (hunk 6 above)
Comment (MatthiasKillat) [reply doc/website/advanced/best-practice-for-testing.md:71]: Yes. We do want to avoid crashing tests even if the sut is defective and should prefer graceful failure. In more exotic cases also endless loops could occur due to previous failures, leading to a test that does not terminate. Using ASSERT to prevent this as much as possible should be required in any test.

File: doc/website/advanced/best-practice-for-testing.md [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+    auto number4 = number2 + number3;
+    auto number5 = number2 - number3;
+
+    EXPECT_TRUE(number4 == 5U);
+    EXPECT_EQ(1U, static_cast<uint64_t>(number5));
+}
+```
+
+Done and we reached 100% coverage. We can clap ourselves on the shoulders and move on. Well, except that we can't.
+
```
Comment (MatthiasKillat) [inline doc/website/advanced/best-practice-for-testing.md:141]: meets the eye

File: doc/website/advanced/best-practice-for-testing.md [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+The test above has several major and minor flaws. The first thing that leaps to the eye, it doesn't follow the **AAA** pattern. When the test fails, we don't know why and have to look in the code to figure out what went wrong. But even if only one aspect of the class would have been tested, there would still be essentially just a `[  FAILED  ] SingleDigitNumber_test.TestClass` output and we would need to look at the code to know what exactly is broken. This test case checks too many aspects of the code and doesn't have a sensible name.
+
```
Comment (MatthiasKillat) [inline doc/website/advanced/best-practice-for-testing.md:143]: The test is bad, no doubt, but we usually want to look at the name to get a rough idea why it failed and then we need to look which EXPECT failed. It is often infeasable to just have one EXPECT per test (e.g. an element is there so the container size is 1 and it is the element we put into the container and others). In other words: there must be some flexibility in what is "too many".

File: doc/website/advanced/best-practice-for-testing.md [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+Here, ZOMBIES comes into play and gives us some guide to identify this cases.
+
+Some of this test might made it necessary to define the behavior of the class in the first place, e.g. defining that an `operator+` will saturate to the max value instead of overflowing.
+
+## How To Do It Better
+
+At first, the test is split into multiple test cases according the AAA pattern
+
+```c++
+TEST_F(SingleDigitNumber_test, DefaultConstructedObjectIsCorrectlyInitialized)
```
Comment (MatthiasKillat) [inline doc/website/advanced/best-practice-for-testing.md:163]: If there is no complex initialization needed and the sut initialization is not universal to each tests we could also use TEST without fixtures. Probably not a good idea to advocate though, we can do useful things in the fixture like suppressing output by capturing ``std::cout`` and so on.

It should almost always be preferred to use the fixture, as you demonstrate (but tests without should be still allowed).

File: doc/website/advanced/best-practice-for-testing.md [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+```
+
+These are some examples showing how to apply the ZOMBIES principle to find good test cases.
+They also exert all the good practices mentioned previously, like clear and distinct names or avoiding magic numbers.
+
+# Slightly More Advanced Topics
+
+## Typed Tests
+
+There are situations when test cases only vary in the type applied to the `sut`. In this case typed tests can be used to reduce repetition.
```
Comment (MatthiasKillat) [inline doc/website/advanced/best-practice-for-testing.md:232]: This can also be used to vary additional parameters in a kind of test configuration consisting of the SUT type and those parameters (such as duration of a test, number of elements of a container and so on). But this is mainly for fairly complex tests and probably does not belong into a beginner`s guideline document.

File: doc/website/advanced/best-practice-for-testing.md (hunk 1 above)
Comment (MatthiasKillat) [reply doc/website/advanced/best-practice-for-testing.md:257]: I am with @elBoberido here. Sometimes it is not easy to express the test intention with those Suffixes and the author needs some wiggle room.

**Review by MatthiasKillat — DISMISSED** [review]

File: doc/website/advanced/best-practice-for-testing.md (hunk 1 above)
Comment (elfenpiff) [reply doc/website/advanced/best-practice-for-testing.md:257]: I am with @marthtz  here. Let's say you are the poor developer who has to fix a test named like `CallingMethodWithParameter`. Then you now what is tested here but you do not know what kind of outcome the author expected. Should this always work, should it fail?

Then you take a look inside the code see that it fails and the implementation fails but was it really the intention here. Maybe two things came together - an implementation error in the test and in the actual implementation?!

In the end I feel that test names are like tiny requirements which states DoingBlaLeadsToBehaviorFuu.

File: doc/website/advanced/best-practice-for-testing.md (hunk 14 above)
Comment (elBoberido) [reply doc/website/advanced/best-practice-for-testing.md:143]: I'll change `aspects` to `state transistions` and note that all side effects of that transition must be checked

**Review by dkroenke — APPROVED** [review]

File: doc/website/advanced/best-practice-for-testing.md [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+- overflow with `operator+`, e.g. adding `7` and `8`
+- underflow with `operator-`, e.g. subtracting `8` from `7`
+
+Here, ZOMBIES comes into play and gives us some guide to identify this cases.
+
+Some of this test might made it necessary to define the behavior of the class in the first place, e.g. defining that an `operator+` will saturate to the max value instead of overflowing.
+
+## How To Do It Better
+
+At first, the test is split into multiple test cases according the AAA pattern
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:158]: according to

File: doc/website/advanced/best-practice-for-testing.md [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+## Typed Tests
+
+There are situations when test cases only vary in the type applied to the `sut`. In this case typed tests can be used to reduce repetition.
+The there is a section for [typed tests](https://github.com/google/googletest/blob/master/docs/advanced.md#typed-tests) in the advanced gtest documentation.
+
+A more thorough [example](https://github.com/google/googletest/blob/release-1.8.1/googletest/samples/sample6_unittest.cc) is in the gtest github repository.
+
+## Parameterized Tests
+
+Similar to typed tests, there are cases where the same test case should run with multiple parameter.
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:237]: parameter**s**

File: doc/website/advanced/best-practice-for-testing.md [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+One example is the conversion of `enum` values to strings. While this can be done in a loop, a better approach are parameterized test.
+
+[This](https://www.sandordargo.com/blog/2019/04/24/parameterized-testing-with-gtest) is quite a good blog post to get into parameterized tests. Additionally, there is the a section in the advanced [documentation](https://github.com/google/googletest/blob/master/docs/advanced.md#value-parameterized-tests) for gtest.
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:240]: ``Additionally, there is a section...`` (remove ``the``)

File: doc/website/advanced/best-practice-for-testing.md [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+
+The block post mentions tuples to pass multiple parameter at once. Since tuples can become cumbersome to use, especially if parameters are rearranged, it is recommended to create a `struct` to wrap the parameters instead.
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/advanced/best-practice-for-testing.md:242]: parameter**s**

**Review by dkroenke — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]
